src/operations/blas3/gemm_no_local.hpp

```diff
@@ -394,76 +415,444 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
+   * major instead of column.
+   * @tparam PointerType: the type of the input matrix.
+   * @tparam RowCheckType: the type of a function used for checking the
+   * boundary in the row direction for blocks of data located at the edge of the
+   * input matrix.
+   * @tparam ColCheckType: the type of a function used for checking the
+   * boundary in the column direction for blocks of data located at the edge of
+   * the input matrix.
+   * @param ptr : the input matrix A.
+   * @param reg : the private register for A.
+   * @param ptr_next: offset for the next value to be loaded.
    * @param ld : the leading dimension of the input matrix.
-   * @param index: the start position of the block of data to be loaded.
-   * @param chk_boundary: an instance of the check_boundary function
+   * @param in_row : function which checks the boundary in the row direction.
+   * @param in_col : function which checks the boundary in the col direction.
+   * @param out_of_range: exits the function early if block is out of range.
    */
 
-  template <index_t item_size, index_t next_element, bool check_block,
-            typename PointerType, typename check_boundary>
-  SYCL_BLAS_INLINE void load(PointerType ptr, element_t *reg, const index_t &ld,
-                             index_t index, const check_boundary &chk_boundary,
-                             const bool out_of_range) noexcept {
+  template <index_t rows, index_t cols, index_t next_element, bool check_row,
+            bool check_col, index_t work_per_load, bool trans,
+            typename PointerType, typename RowCheckType, typename ColCheckType>
+  SYCL_BLAS_INLINE typename std::enable_if<!trans>::type load_block_a(
+      PointerType ptr, element_t *reg, const index_t &ptr_next,
+      const index_t &ld, const RowCheckType &in_row, const ColCheckType &in_col,
+      const bool out_of_range) noexcept {
     if (out_of_range) {
       return;
     }
 #pragma unroll
-    for (int i = 0; i < item_size; i++) {
-      reg[i] =
-          do_check<check_block>(chk_boundary(index)) ? ptr[0] : element_t(0);
+    for (int i = 0; i < cols; i++) {
+#pragma unroll
+      for (int j = 0; j < rows / work_per_load; j++) {
+        // Check that the last element of the packet loaded is in range
+        bool in_range = do_check<check_row>(in_row(work_per_load - 1)) &&
+                        do_check<check_col>(in_col(i));
+
+        cl::sycl::vec<element_t, work_per_load> in_vec{0};
+        if (in_range) {
+          // if in range perform a vectorised load
+          in_vec.template load<address_t::global_space>(0, ptr + j * ptr_next);
+        } else {
+          // if not in range perform element-wise load checking boundaries at
+          // each load.
+#pragma unroll
+          for (int l = 0; l < work_per_load; l++) {
+            if (do_check<check_row>(in_row(l)) &&
+                do_check<check_col>(in_col(i))) {
+              reinterpret_cast<element_t *>(&in_vec)[l] =
```

Comment:
I'm not 100% sure this is safe. Unfortunately SYCL doesn't provide a programmatic way to set vector elements. I don't think this should block this PR, but we could think about adding something like the helpers used in SYCL-DNN for this: https://github.com/codeplaysoftware/SYCL-DNN/blob/master/src/helpers/vector_element.h

Reply:
I agree it probably needs further discussion. Helpers could work assuming there is no performance penalty.